app/src/main/java/com/cmput301f20t21/bookfriends/services/RequestService.java

```diff
@@ -0,0 +1,95 @@
+package com.cmput301f20t21.bookfriends.services;
+
+import com.cmput301f20t21.bookfriends.enums.REQUEST_STATUS;
+import com.google.android.gms.tasks.OnSuccessListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.firestore.CollectionReference;
+import com.google.firebase.firestore.DocumentReference;
+import com.google.firebase.firestore.FirebaseFirestore;
+import com.google.firebase.firestore.QuerySnapshot;
+import com.google.firebase.firestore.WriteBatch;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+
+public class RequestService {
+    private CollectionReference requestCollection;
+
+    private static final RequestService instance = new RequestService();
+
+    private RequestService() {
+        requestCollection = FirebaseFirestore.getInstance().collection("requests");
+    }
+
+    public static RequestService getInstance() {
+        return instance;
+    }
+
+    /**
+     * get request by bookId
+     * @param bookId
+     * @return Task QuerySnapshot for request
+     */
+    public Task<QuerySnapshot> getByBookId(String bookId) {
+        return requestCollection.whereEqualTo("bookId", bookId).get();
+    }
+
+    /**
+     * add request by bookId and requesterId
+     * @param bookId, requesterId
+     * @return Task DocumentReference for request
+     */
+    public Task<DocumentReference> add(String bookId, String requesterId) {
+        HashMap<String, String> data = new HashMap<>();
+        data.put("bookId", bookId);
+        data.put("requesterId", requesterId);
+        data.put("status", REQUEST_STATUS.OPENED.toString());
+        return requestCollection.add(data);
+    }
+
+    /**
+     * accept a request
+     * @param id
+     * @return Task Void
+     */
+    public Task<Void> accept(String id) {
+        return requestCollection.document(id).update("status", REQUEST_STATUS.ACCEPTED.toString());
+    }
+
+    /**
+     * deny a request
+     * @param id
+     * @return Task Void
+     */
+    public Task<Void> deny(String id) {
+        return requestCollection.document(id).update("status", REQUEST_STATUS.DENIED.toString());
+    }
+
+    /**
+     * batch deny requests
+     * @param ids
+     * @return Task Void
+     */
+    public Task<Void> batchDeny(List<String> ids) {
+        WriteBatch batch = FirebaseFirestore.getInstance().batch();
+        for (String id: ids) {
+            DocumentReference request = requestCollection.document(id);
+            batch.update(request, "status", REQUEST_STATUS.DENIED.toString());
+        }
+        return batch.commit();
+    }
+
+    /**
+     * get request that unavailable.
+     * unavailable requests have status: ACCEPTED, BORROWED
+     * @param bookId
+     * @return Task QuerySnapshot for unavailable request
+     */
+    public Task<QuerySnapshot> getUnavailableRequest(String bookId) {
```

Comment:
I'm thinking when we get book, we can fill in the book status by check for requests that are accepted or borrowed. If this query return empty, then book is available. If we implement the system correctly, this query should return only 1 document. We can set the book status to the request status.

Reply:
This could work. But in comparison, what if we store book status as a Book attribute, update them as we update requests if necessary? if we did it right, it should save a lot of query time, and separate book status from requests'.

consider when we fetch all borrowed books: the method you mentioned might query in time (#of requests  * # of books )

Reply:
by US 01.03.01:
As an owner or borrower, I want a book to have a status to be one of: available, requested, accepted, or borrowed.

I think book does need to have a status attribute